### ISSUE TITLE
Project a VectorCoefficient on a GridFunction with a vdim > 3

### ISSUE THREAD

**Issue by CorporalClegg** [mfem/mfem#116]
Hello everyone,

I want to project this Coefficient:

```c++
    class StrainVectorCoefficient: public mfem::VectorCoefficient {
    private:
        mfem::GridFunction &u ;

    public:
        StrainVectorCoefficient(
            mfem::GridFunction &_u )
            : u( _u ), VectorCoefficient(6)
        {
        }
        virtual void Eval(
            mfem::Vector& V,
            mfem::ElementTransformation &T,
            const mfem::IntegrationPoint &ip )
        {
            mfem::DenseMatrix eps ;
            u.GetVectorGradient( T, eps ) ; 
            eps.Symmetrize() ;
            V[0] = eps( 0, 0 ) ;
            V[1] = eps( 1, 1 ) ;
            V[2] = eps( 2, 2 ) ;
            V[3] = eps( 0, 1 ) ;
            V[4] = eps( 0, 2 ) ;
            V[5] = eps( 1, 2 ) ;
        }
        virtual void Read( std::istream &in )
        {
        }
        virtual ~StrainVectorCoefficient()
        {
        }
    } ;
```

on a GridFunction defined like this:
```c++
mfem::FiniteElementCollection * fec = new mfem::L2_FECollection( 2, 3 ) ;
mfem::FiniteElementSpace * pp_fespace = new mfem::FiniteElementSpace( mfem_mesh_ pp_fec, 6,mfem::Ordering::byVDIM ) ;
StrainVectorCoefficient strain_coeff( *x_) ;
mfem::GridFunction * strain = new mfem::GridFunction( pp_fespace ) ;
strain->ProjectCoefficient(strain_coeff) ;
```

but I fall in this assert : 
```
Assertion failed: (vc.GetVDim() <= 3) is false: 
 ... in function: virtual void mfem::NodalFiniteElement::Project(mfem::VectorCoefficient&, mfem::ElementTransformation&, mfem::Vector&) const
 ... at line 257 of file: fem/fe.cpp
```
I work in 3D.

Am I doing something wrong ? If I initialize my StrainVectorCoefficient with a vdim of 3, I am able to retrieve the three fist components of the strain. But I can't retrieve more. Should I increase the dimension of the FiniteElementCollection ?

Thank you in advance !

**Comment by mlstowell** [comment]
Hello,

I don't think you are necessarily doing anything wrong (although I do see what look like cut and paste errors in the GridFunction definition e.g. pp_fec is not declared).  The L2 finite elements are derived from the NodalFiniteElement class and this class does expect that vdim will be no larger than 3.  It makes this assumption so that it can declare a buffer on the stack and avoid allocating temporary memory on the heap every time the Project method is called.

A quick fix would be to increase the size of the buffer used in the project method (this is defined near line 259 of fe.cpp in the method NodalFiniteElement::Project).  You'll also need to alter the MFEM_ASSERT to reflect the new buffer size.

A more general fix would be to create a mutable Vector as a member of NodalFiniteElement that can be sized appropriately and used as a temporary vector in this project method.

Thanks,
Mark

**Comment by CorporalClegg** [comment]
Hi Mark,

Thank you for your answer.
Yes I have made some mistake during the copy/paste :smile: 

So I need to modify the mfem code, I will see (for the moment I work with vector of GridFunction in my own code). maybe I can do a pull request with your suggestion ?

Thank you again

Antoine

**Comment by tzanio** [comment]
Hi Antoine,

Were able to make progress on this issue? Do you need any additional help?

Cheers,
Tzanio

**Comment by CorporalClegg** [comment]
Hi Tzanio,
Thank you I have well understand the problem and I solve it using a dirty hack (thanks Mark :D)

Cheers,
Antoine

**Comment by v-dobrev** [comment]
Fixed in PR #163.
